src/category_theory/connected_components.lean

```diff
@@ -0,0 +1,153 @@
+-/
+
+universes v₁ v₂ v₃ u₁ u₂
+
+noncomputable theory
+
+open category_theory.category
+
+namespace category_theory
+
+attribute [instance, priority 100] is_connected.is_nonempty
+
+variables {J : Type u₁} [category.{v₁} J]
+variables {C : Type u₂} [category.{u₁} C]
+
+/-- This type indexes the connected components of the category `J`. -/
+def connected_components (J : Type u₁) [category.{v₁} J] : Type u₁ := quotient (zigzag.setoid J)
+
+instance [inhabited J] : inhabited (connected_components J) := ⟨quotient.mk' (default J)⟩
+
+/-- Given an index for a connected component, produce the actual component as a full subcategory. -/
+@[derive category]
+def component (j : connected_components J) : Type u₁ := {k : J // quotient.mk' k = j}
+
+/-- The inclusion functor from a connected component to the whole category. -/
+@[derive [full, faithful], simps {rhs_md := semireducible}]
+def include_component (j) : component j ⥤ J :=
+full_subcategory_inclusion _
+
+/-- Each connected component of the category is nonempty. -/
+instance (j : connected_components J) : nonempty (component j) :=
+begin
+  apply quotient.induction_on' j,
+  intro k,
+  refine ⟨⟨k, rfl⟩⟩,
+end
+
+instance (j : connected_components J) : inhabited (component j) := classical.inhabited_of_nonempty'
+
+/-- Each connected component of the category is connected. -/
+instance (j : connected_components J) : is_connected (component j) :=
+begin
+  -- Show it's connected by constructing a zigzag (in `component j`) between any two objects
+  apply is_connected_of_zigzag,
+  rintro ⟨j₁, hj₁⟩ ⟨j₂, rfl⟩,
+  -- We know that the underlying objects j₁ j₂ have some zigzag between them in `J`
+  have h₁₂ : zigzag j₁ j₂ := quotient.exact' hj₁,
+  -- Get an explicit zigzag as a list
+  rcases list.exists_chain_of_relation_refl_trans_gen h₁₂ with ⟨l, hl₁, hl₂⟩,
+  -- Everything which has a zigzag to j₂ can be lifted to the same component as `j₂`.
+  let f : Π x, zigzag x j₂ → component (quotient.mk' j₂) := λ x h, ⟨x, quotient.sound' h⟩,
+  -- Everything in our chosen zigzag from `j₁` to `j₂` has a zigzag to `j₂`.
+  have hf : ∀ (a : J), a ∈ l → zigzag a j₂,
+  { intros i hi,
+    apply list.chain.induction (λ t, zigzag t j₂) _ hl₁ hl₂ _ _ _ (or.inr hi),
+    { intros j k,
+      apply relation.refl_trans_gen.head },
+    { apply relation.refl_trans_gen.refl } },
+  -- Now lift the zigzag from `j₁` to `j₂` in `J` to the same thing in `component j`.
+  refine ⟨l.pmap f hf, _, _⟩,
+  { refine @@list.chain_pmap_of_chain _ _ _ f (λ x y _ _ h, _) hl₁ h₁₂ _,
+    exact zag_of_zag_obj (include_component _) h },
+  { erw list.last_pmap _ f (j₁ :: l) (by simpa [h₁₂] using hf) (list.cons_ne_nil _ _),
+    exact subtype.ext hl₂ },
+end
+
+/--
+The disjoint union of `J`s connected components, written explicitly as a sigma-type with the
+category structure.
+This category is equivalent to `J`.
+-/
+abbreviation decomposed (J : Type u₁) [category.{v₁} J] :=
```

Comment:
This name seems somewhat generic, and it's in the `category_theory` namespace.
Same holds for `inclusion` below it, which is even more generic, maybe.

Maybe they should be namespaced?

Reply:
Sounds like a good idea - or just picking a more descriptive name for these. Do you have suggestions?

Reply:
I couldn't really come up with a good name for this one... you could ask on Zulip, maybe someone has inspiration